aggregator/map.go

```diff
@@ -123,88 +145,62 @@ func (m *metricMap) Tick(target time.Duration) tickResult {
 	}
 }
 
-func (m *metricMap) Close() {
-	m.metricLists.Close()
-}
-
-func (m *metricMap) deleteExpired(target time.Duration) int {
-	now := m.nowFn()
-
-	// Determine batch size.
-	m.RLock()
-	elemsLen := m.entryList.Len()
-	if elemsLen == 0 {
-		// If the list is empty, nothing to do.
-		m.RUnlock()
-		return 0
-	}
-	batchSize := int(math.Max(1.0, math.Ceil(m.batchPercent*float64(elemsLen))))
-	numBatches := int(math.Ceil(float64(elemsLen) / float64(batchSize)))
-	targetPerBatch := target / time.Duration(numBatches)
-	currElem := m.entryList.Front()
-	m.RUnlock()
-
-	// NB(xichen): if this runs frequently enough, stash the expired buffer
-	// in the map object and reuse the buffer.
-	var (
-		numExpired int
-		expired    = make([]hashedEntry, 0, batchSize)
-		batchIdx   int
-	)
-	for currElem != nil {
-		m.RLock()
-		for numChecked := 0; numChecked < batchSize && currElem != nil; numChecked++ {
-			nextElem := currElem.Next()
-			hashedEntry := currElem.Value.(hashedEntry)
-			if hashedEntry.entry.ShouldExpire(now) {
-				expired = append(expired, hashedEntry)
-			}
-			currElem = nextElem
-		}
-		m.RUnlock()
+func (m *metricMap) SetRuntimeOptions(opts runtime.Options) {
+	m.Lock()
+	m.runtimeOpts = opts
+	m.Unlock()
 
-		// Actually purging the expired entries.
-		if len(expired) >= batchSize {
-			numExpired += m.purgeExpired(now, expired)
-			for i := range expired {
-				expired[i] = emptyHashedEntry
-			}
-			expired = expired[:0]
-		}
+	// NB(xichen): we hold onto the entry list deletion lock here to ensure no
+	// entries get deleted while we iterate over the list, otherwise we may update
```

Comment:
What happens if we update the expired entries? aren't they gonna be deleted later anyway? 
It would be much better to not having to have a delete lock here, using the 2 locks adds complexity and make the code prone to dead locks.

Reply:
I'd argue deleting expired entries is more dangerous. Expired entries are put back into the object pool (or reused), so we'd be examining / mutating state of an object we should not even have access to because it's already expired.

There are only 2 places where the deletion lock is used, and as long as the locks are acquired in the same order, we can ensure no deadlocks are going to happen. This is primarily needed for the sake of performance so that entries can be inserted while the runtime options are updated.

Reply:
Can this be replace by doing a entry.ShouldExpire check before SetRuntimeOptions()?

Reply:
The issue is we don't know when an entry is expired or not without the lock, the entry being examined could be expired from the ticking goroutine from underneath us, and we'd be calling `ShouldExpire` on the entry that's already expired.